07 post tag

hidekazu-hhh opened the pull request:
## 投稿すると他のユーザーの投稿に紐づいたタグが消えてしまう

投稿した学習内容とタグを記録して学習記録をカレンダーに表示させるアプリを制作しています。

## 起きているエラーについて

ユーザーA が　タグ　（プログラミング）を登録してタグの詳細画面に遷移すると

ユーザーBの投稿と紐づいていたタグ（プログラミング）が消えてしまう。

タグに紐づいた自分の投稿のみを一覧で表示させるとエラーが起きますが、タグに紐づいた全ユーザーの投稿を表示させた場合、このエラーは発生しません。

タグの詳細画面にタグに紐づいた自分の投稿のみを表示させ、他のユーザーの投稿に紐づいたタグが消えないようにしたいのですが原因の特定と対処の仕方がわからず、行き詰まっています。

## 該当のコード

TagsController

```

def show
    user = User.find_by(id: current_user.id)
    @tag = user.tags.find(params[:id])
    @tag.posts = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")
    # タグに紐づいた現在のユーザーのみの投稿
  end　　
```

Tags/show.html.erb

```
<div class="col-md-12">
<div class="card border-secondary mb-3" style="max-width: 30rem;">
  <% @tag.posts.each do |post| %>
    <%#タグに紐づく投稿の全件取得%>
 <div class="card-header">ユーザ名〇〇</div>
  <div class="card-body text-dark">
    <p class="card-text"><%= post.content%></p>
   <br>
    <%= link_to '編集', post_path(post) %>
    <%= link_to '削除', post_path(post), method: :delete, data: {confirm: '削除しますか？'} %>
    <br>
    <br>
     <%= l post.created_at, format: :long %>
    <br>
  </div>
   <% end %>
  </div>
</div>
```

## ログ

```

Post Create (0.3ms)  INSERT INTO `posts` (`content`, `start_time`, `user_id`, `created_at`, `updated_at`) VALUES ('川口太郎の\nプログラミング', '2022-06-06 00:00:00', 5, '2022-06-05 15:07:05.237485', '2022-06-05 15:07:05.237485')
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (0.3ms)  COMMIT
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (0.4ms)  SELECT `tags`.`name` FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 70
  ↳ app/models/post.rb:15:in `save_tags'
[]
  Tag Load (0.4ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`name` = 'プログラミング' LIMIT 1
  ↳ app/models/post.rb:39:in `block in save_tags'
   (0.3ms)  BEGIN
  ↳ app/models/post.rb:40:in `block in save_tags'
  Tag Load (1.2ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`id` = 16 LIMIT 1
  ↳ app/models/post.rb:40:in `block in save_tags'
  User Load (0.3ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  ↳ app/models/post.rb:40:in `block in save_tags'

中間テーブル
  **PostTag Create** (0.3ms)  INSERT INTO `post_tags` (`post_id`, `tag_id`, `user_id`, `created_at`, `updated_at`) VALUES (70, 16, 5, '2022-06-05 15:07:05.256623', '2022-06-05 15:07:05.256623')
  ↳ app/models/post.rb:40:in `block in save_tags'
   (0.3ms)  COMMIT
  ↳ app/models/post.rb:40:in `block in save_tags'
Redirected to http://localhost:3000/posts
Completed 200 OK in 26ms (ActiveRecord: 4.6ms | Allocations: 12032)

Started GET "/posts" for ::1 at 2022-06-05 15:07:05 +0900
Processing by PostsController#index as HTML
  User Load (0.4ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  Rendering posts/index.html.erb within layouts/application
  Tag Load (1.1ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`user_id` = 5
  ↳ app/views/posts/index.html.erb:9
   (0.4ms)  SELECT COUNT(*) FROM `posts` INNER JOIN `post_tags` ON `posts`.`id` = `post_tags`.`post_id` WHERE `post_tags`.`tag_id` = 16
  ↳ app/views/posts/index.html.erb:10
  Post Load (0.6ms)  SELECT `posts`.* FROM `posts` WHERE `posts`.`user_id` = 5 ORDER BY `posts`.`created_at` DESC
  ↳ app/views/posts/index.html.erb:15
  Tag Load (0.4ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 70
  ↳ app/views/posts/index.html.erb:28:in `map'
  Rendered simple_calendar/_month_calendar.html.erb (Duration: 11.8ms | Allocations: 5677)
  Rendered posts/index.html.erb within layouts/application (Duration: 20.2ms | Allocations: 8592)
[Webpacker] Everything's up-to-date. Nothing to do
  Rendered shared/_header.html.erb (Duration: 0.5ms | Allocations: 206)
  Rendered shared/_flash_message.html.erb (Duration: 0.1ms | Allocations: 17)
  Rendered shared/_footer.html.erb (Duration: 0.2ms | Allocations: 5)
Completed 200 OK in 33ms (Views: 26.6ms | ActiveRecord: 2.9ms | Allocations: 13715)

Started GET "/posts/70" for ::1 at 2022-06-05 15:08:09 +0900
Processing by PostsController#show as HTML
  Parameters: {"id"=>"70"}
  User Load (0.3ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  Post Load (1.9ms)  SELECT `posts`.* FROM `posts` WHERE `posts`.`user_id` = 5 AND `posts`.`id` = 70 LIMIT 1
  ↳ app/controllers/posts_controller.rb:8:in `show'
  Rendering posts/show.html.erb within layouts/application
  Tag Load (0.5ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 70
  ↳ app/views/posts/show.html.erb:19
  Rendered posts/show.html.erb within layouts/application (Duration: 4.3ms | Allocations: 1401)
[Webpacker] Everything's up-to-date. Nothing to do
  Rendered shared/_header.html.erb (Duration: 0.3ms | Allocations: 207)
  Rendered shared/_flash_message.html.erb (Duration: 0.1ms | Allocations: 17)
  Rendered shared/_footer.html.erb (Duration: 0.0ms | Allocations: 5)
Completed 200 OK in 19ms (Views: 10.4ms | ActiveRecord: 2.8ms | Allocations: 7000)

投稿一覧画面
Started GET "/posts" for ::1 at 2022-06-05 15:08:13 +0900
Processing by PostsController#index as HTML
  User Load (0.4ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  Rendering posts/index.html.erb within layouts/application
  Tag Load (2.3ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`user_id` = 5
  ↳ app/views/posts/index.html.erb:9
   (0.6ms)  SELECT COUNT(*) FROM `posts` INNER JOIN `post_tags` ON `posts`.`id` = `post_tags`.`post_id` WHERE `post_tags`.`tag_id` = 16
  ↳ app/views/posts/index.html.erb:10
  Post Load (0.9ms)  SELECT `posts`.* FROM `posts` WHERE `posts`.`user_id` = 5 ORDER BY `posts`.`created_at` DESC
  ↳ app/views/posts/index.html.erb:15
  Tag Load (1.0ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 70
  ↳ app/views/posts/index.html.erb:28:in `map'
  Rendered simple_calendar/_month_calendar.html.erb (Duration: 10.4ms | Allocations: 5677)
  Rendered posts/index.html.erb within layouts/application (Duration: 21.6ms | Allocations: 8586)
[Webpacker] Everything's up-to-date. Nothing to do
  Rendered shared/_header.html.erb (Duration: 0.6ms | Allocations: 206)
  Rendered shared/_flash_message.html.erb (Duration: 0.1ms | Allocations: 17)
  Rendered shared/_footer.html.erb (Duration: 0.0ms | Allocations: 5)
Completed 200 OK in 36ms (Views: 27.5ms | ActiveRecord: 5.1ms | Allocations: 13717)

タグの詳細画面
Started GET "/tags/16" for ::1 at 2022-06-05 15:09:36 +0900
Processing by TagsController#show as HTML
  Parameters: {"id"=>"16"}
  User Load (0.5ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  User Load (1.7ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5 LIMIT 1
  ↳ app/controllers/tags_controller.rb:8:in `show'
  Tag Load (1.2ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`user_id` = 5 AND `tags`.`id` = 16 LIMIT 1
  ↳ app/controllers/tags_controller.rb:9:in `show'
  Post Load (0.4ms)  SELECT `posts`.* FROM `posts` INNER JOIN `post_tags` ON `posts`.`id` = `post_tags`.`post_id` WHERE `post_tags`.`tag_id` = 16 AND `posts`.`user_id` = 5 ORDER BY created_at DESC
  ↳ app/controllers/tags_controller.rb:10:in `show'
  User Load (0.8ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 5
  ↳ app/controllers/tags_controller.rb:10:in `show'
  Post Load (0.4ms)  SELECT `posts`.* FROM `posts` INNER JOIN `post_tags` ON `posts`.`id` = `post_tags`.`post_id` WHERE `post_tags`.`tag_id` = 16
  ↳ app/controllers/tags_controller.rb:10:in `show'
   (0.2ms)  BEGIN

タグの詳細画面に遷移すると他のユーザーの投稿に紐づいたタグが消えてしまう。
  ↳ app/controllers/tags_controller.rb:10:in `show'
  **PostTag Destroy (1.2ms)  DELETE FROM `post_tags` WHERE `post_tags`.`tag_id` = 16 AND `post_tags`.`post_id` = 69**
  ↳ app/controllers/tags_controller.rb:10:in `show'
   (0.4ms)  COMMIT
  ↳ app/controllers/tags_controller.rb:10:in `show'
  Rendering tags/show.html.erb within layouts/application
  Rendered tags/show.html.erb within layouts/application (Duration: 0.5ms | Allocations: 247)
[Webpacker] Everything's up-to-date. Nothing to do
  Rendered shared/_header.html.erb (Duration: 0.3ms | Allocations: 207)
  Rendered shared/_flash_message.html.erb (Duration: 0.0ms | Allocations: 17)
  Rendered shared/_footer.html.erb (Duration: 0.0ms | Allocations: 5)
Completed 200 OK in 31ms (Views: 6.0ms | ActiveRecord: 6.8ms | Allocations: 12393)
```

考えられる原因

タグの詳細画面にはタグに紐づいた自分の投稿のみを表示させていますが、現状のTags show　の記述ではタグの詳細画面に遷移する時に整合性を保つため、他のユーザーに紐づいた投稿のタグが削除されてしまう。

タグの詳細画面を全ユーザーの投稿にすると今回のエラーは起きません。

Tagsのshowを以下に戻す。 

```
def show    
    @tag = Tag.find(params[:id])
end　　
```

kerochelo commented:
@hidekazu-hhh 
下記のログが怪しいですね！

```
 ↳ app/controllers/tags_controller.rb:10:in `show'
  **PostTag Destroy (1.2ms)  DELETE FROM `post_tags` WHERE `post_tags`.`tag_id` = 16 AND `post_tags`.`post_id` = 69**
  ↳ app/controllers/tags_controller.rb:10:in `show'
   (0.4ms)  COMMIT
  ↳ app/controllers/tags_controller.rb:10:in `show'
```

該当するtags_controller.rbの10行目の処理である下記の部分はコードは処理の意図的にあっていますでしょうか？
こちらを確認すれば直りそうです！
```
@tag.posts = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")
```

hidekazu-hhh commented:
> 該当するtags_controller.rbの10行目の処理である下記の部分はコードは処理の意図的にあっていますでしょうか？
> こちらを確認すれば直りそうです！
@tag.posts = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")


ユーザー自身のタグに紐づいた投稿一覧のみを表示させたいのですが上手く行きません。
上記のコードで条件に該当する検索をかけたいのですが、なぜか他のユーザーに紐づいていた投稿のタグが削除されてしまう原因がよくわかっていません。アソシエーションが影響しているのでしょうか？

kerochelo commented:
@hidekazu-hhh 

`@tag.posts`に上書きしちゃっている部分はあっていますかね？
記載されているコード流れを見ると
```ruby
# current_userの情報取得
user = User.find_by(id: current_user.id)
# current_userが登録しているタグ情報を取得
@tag = user.tags.find(params[:id])
# @tagに紐づいているposts情報を@tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")で上書きしている
# 要するに@tag.postsを削除して@tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")を入れている
@tag.posts = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")
```
の流れになっていますね

タグに紐づいた情報一覧を取得するってことはアソシエーションを利用する必要があるというのは理解しているかと思います！
おそらくしたいこととしては
```ruby
@tag_posts = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")
```
のように新たな変数に情報を入れたいのかなとー
この方法でもいいですが、
```ruby
@posts = current_user....
```
のように１行で取れそうですね

kerochelo commented:
@hidekazu-hhh 

あと返答時には
@でメンション入れていただけると助かります:pray:

hidekazu-hhh commented:
@kerochelo 

**エラーの原因**
今回、同じ変数名で定義していたのでコードが上書きされていた。
@tag.postsは元々、アソシエーションで設定されている変数。

> ```ruby
> @posts = current_user....
> ```
> 
> のように１行で取れそうですね


コード修正しました。
```

@tag = Tag.find(params[:id])
@tag_user_posts  = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")
```
新たに変数名を定義。
タグに紐づいた投稿を全件取得。
whereで現在のユーザーを取得。








他に一行で取れる方法が思いつきませんでした。

hidekazu-hhh commented:
# Unpermitted parameter: :tagの原因の特定と解消がしたい。

以前、質問した件についてエラーが解消した旨を返答し現在、回答待ちですが、

別件で質問があり投稿しました。よろしくお願いします。

投稿時に、表題のエラーが発生。

post と　tag は　多対多　として中間テーブルを設けています。

一つの投稿フォームにタグも含め登録し、

送信されたpostのパラメーターには　tag が含まれています。

```
post
<%= post.label :content, '学習内容' %><br>
<%= post.text_area :content %><br>
<%= post.label :start_time, '学習日' %><br>
<%= post.date_select :start_time %><br>

<%= post.label :tag, 'タグ' %><br>
<%= post.text_field :tag, value: @post.tags.pluck(:name) %><br>
<%= post.submit %>
```

ログ

```
 Processing by PostsController#create as JS
Parameters: 
{"authenticity_token"=>"CBd/N70btgHOxmrFOeuKp12ni2kG/KBVull4jOLOhKKWgbgaE6aHykZG5gg4sluRYnP6zYNUascbKRwtjhMxYA==
", "post"=>{"content"=>"あどみん太郎", "start_time(1i)"=>"2022", "start_time(2i)"=>"6", "start_time(3i)"=>"11", **"tag"=>"プログラミング"**}, "commit"=>"登録する"}
```

タグの登録などに関するロジックは　`post.rb`　に記述しています。　 

`save_tags` メソッドで入力フォームから送信されたタグが　

元々と登録しているタグでなければ新たにタグを作成。

中間テーブルへ登録。

```
class Post < ApplicationRecord
  belongs_to :user
  has_many :post_tags, dependent: :destroy
  has_many :tags, through: :post_tags

  validates :content, presence: true, length: { maximum:65535}

  def save_tags(tags)

    tag_list = tags.split(/[[:blank:]]+/)

    # 自分自身に関連づいたタグを取得する
    current_tags = self.tags.pluck(:name)

    # (1) 元々自分に紐付いていたタグと投稿されたタグの差分を抽出
    #   -- 記事更新時に削除されたタグ
    old_tags = current_tags - tag_list

    # (2) 投稿されたタグと元々自分に紐付いていたタグの差分を抽出
    #   -- 新規に追加されたタグ
    new_tags = tag_list - current_tags

    p current_tags

    # tag_mapsテーブルから、(1)のタグを削除
    #   tagsテーブルから該当のタグを探し出して削除する
    old_tags.each do |old|
      # tag_mapsテーブルにあるpost_idとtag_idを削除
      #   後続のfind_byでtag_idを検索
      self.tags.delete Tag.find_by(name: old)
    end

    # tagsテーブルから(2)のタグを探して、tag_mapsテーブルにtag_idを追加する
    new_tags.each do |new|
      # 条件のレコードを初めの1件を取得し1件もなければ作成する
      # find_or_create_by : https://railsdoc.com/page/find_or_create_by
      new_post_tag = Tag.find_or_create_by(name: new)
      post_tags.new(user_id: user_id,tag_id: new_post_tag.id).save
      # tag_mapsテーブルにpost_idとtag_idを保存
      #   配列追加のようにレコードを渡すことで新規レコード作成が可能
      # self.tags << new_post_tag
    end
    
  end
end
```

postsコントローラ

```
class PostsController < ApplicationController
 
  def new
    @post = Post.new
  end

  def create
    @post = current_user.posts.new(post_params)    
    if @post.save
      @post.save_tags(params[:post][:tag])
      redirect_to posts_path
    else
      render :new
    end
  end

  private

  def post_params
    params.require(:post).permit(:content, :start_time)
  end
end
```

## 起きているエラー

Unpermitted parameter: :tag　とエラーが出るがテーブルに保存される。

ログ

```
Use Ctrl-C to stop
Started POST "/posts" for ::1 at 2022-06-12 15:54:46 +0900
   (0.3ms)  SELECT `schema_migrations`.`version` FROM `schema_migrations` ORDER BY `schema_migrations`.`version` ASC
Processing by PostsController#create as JS
  Parameters: {"authenticity_token"=>"SxuTUf0vjE+MxBABdeCauRIuQ7MZt42XTyxsb+45g4bKzDPnH8WcOxOjVa6/UwFBEAv6u8IzHuDY/SZbH4syUA==", "post"=>{"content"=>"Ruby", "start_time(1i)"=>"2022", "start_time(2i)"=>"6", "start_time(3i)"=>"12", "tag"=>"プログラミング"}, "commit"=>"登録する"}
  User Load (0.5ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 6 LIMIT 1
Unpermitted parameter: :tag
   (0.2ms)  BEGIN

  ↳ app/controllers/posts_controller.rb:19:in `create'
  Post Create (0.6ms)  INSERT INTO `posts` (`content`, `start_time`, `user_id`, `created_at`, `updated_at`) VALUES ('Ruby', '2022-06-12 00:00:00', 6, '2022-06-12 15:54:46.389486', '2022-06-12 15:54:46.389486')
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (1.1ms)  COMMIT
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (0.5ms)  SELECT `tags`.`name` FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 151
  ↳ app/models/post.rb:17:in `save_tags'
[]
  Tag Load (1.8ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`name` = 'プログラミング' LIMIT 1
  ↳ app/models/post.rb:41:in `block in save_tags'
   (0.2ms)  BEGIN
  ↳ app/models/post.rb:42:in `block in save_tags'
  Tag Load (0.3ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`id` = 16 LIMIT 1
  ↳ app/models/post.rb:42:in `block in save_tags'
  User Load (0.3ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 6 LIMIT 1
  ↳ app/models/post.rb:42:in `block in save_tags'
  PostTag Create (0.4ms)  INSERT INTO `post_tags` (`post_id`, `tag_id`, `user_id`, `created_at`, `updated_at`) VALUES (151, 16, 6, '2022-06-12 15:54:46.462460', '2022-06-12 15:54:46.462460')
  ↳ app/models/post.rb:42:in `block in save_tags'
   (0.6ms)  COMMIT
  ↳ app/models/post.rb:42:in `block in save_tags'
Redirected to http://localhost:3000/posts
Completed 200 OK in 140ms (ActiveRecord: 11.8ms | Allocations: 46171)
```

## 試したこと

tag を配列で受け取る記述。エラーは変わらず。

```
private

  def post_params
    # params.require(:post).permit(:content, :start_time)
    params.require(:post).permit(:content, :start_time, tag: [])
  end
end
```

ストロングパラメータに :tag　を記述**。**post モデルにはtag という属性がないため。

下記のエラーとなりデータは保存できない。　

```
private

  def post_params
    # params.require(:post).permit(:content, :start_time)
    params.require(:post).permit(:content, :start_time,　：tag)
  end
end
```

```
ActiveModel::UnknownAttributeError (unknown attribute 'tag' for Post.
 raise UnknownAttributeError.new(self, k)
          ^^^^^):
```

多対多　ストロングパラメータで、

検索しましたが本件の事例にあった記事は見当たりませんでした。

デバック

```
[1] pry(#<PostsController>)> params
=> <ActionController::Parameters {"authenticity_token"=>"f9xTXi9WHcc/8zj3X8oAdij+Jf1fRIBivVrogvca6ljhSpRzgessDLdztDpek9FAFypUWdrsSvAcKowjm8dfmg==", "post"=><ActionController::Parameters {"content"=>"あどみん太郎", "start_time(1i)"=>"2022", "start_time(2i)"=>"6", "start_time(3i)"=>"12", "tag"=>"英語"} permitted: false>, "commit"=>"登録する", "controller"=>"posts", "action"=>"create"} permitted: false>

[2] pry(#<PostsController>)> params.require(:post)
=> <ActionController::Parameters {"content"=>"あどみん太郎", "start_time(1i)"=>"2022", "start_time(2i)"=>"6", "start_time(3i)"=>"12", "tag"=>"英語"} permitted: false>
```

## 考えられる原因

params には postをキーとして　tag: が送信されている。posts のcreate時に、 

post.rb で定義した`save_tags`　メソッドを使い`:tag`を保存しているが、ログを見るとカラになっているため、postsコントローラーのcreate アクションの記述が不足している。又はストロングパラメータの記述が誤っている。

ログ

```
↳ app/controllers/posts_controller.rb:19:in `create'
   (0.5ms)  SELECT `tags`.`name` FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 151
  ↳ app/models/post.rb:17:in `save_tags'
[]
```

takyyi commented:
@hidekazu-hhh
```Unpermitted parameter: :tag```が出るのは、PostsController#createにおいて、tagをパラメータとして送っているが、
ストロングパラメータで許可されていないからですね。
ですが、```@post.save_tags(params[:post][:tag])```の方の処理で保存されているのかと思います。
PostとTagは多数多の関係にして管理しているかと思うので、Postテーブルにtagカラムは必要ないかと思います。
うまくtagテーブルで管理できるよう処理を変更してみてください。

hidekazu-hhh commented:
方向性として、
def post_params　には:tag の記述は不要。 
ただ、現状、tag カラムを許可するストロングパラメータがないので、tagモデル用のストロングパラメータを作成する。

posts .new.htenl.erb  にtag のデータも一緒に送信しているので、
post createアクション内に必要なコードを記述する。
でしょうか？　
どのようにしてコードを記述してよいかイメージができていません。

busitora commented:
## tag 一覧取得について

> @tag_user_posts  = @tag.posts.where(user_id: current_user.id).includes(:user).order("created_at DESC")

こちらのコードですが、正しくアソシエーションが設定されているのであれば (user側に)、 current_user.tags (現在のユーザーが投稿したタグ一覧) でとれるはずです。
コントローラー (create) 時に、 デバッグし、 `ls current_user` などして current_userからtags メソッドが生えているか確認し、もしとなければ 
Userモデルを見直してみてください！  (おそらく当該のやりたいことはぐぐるとあるはず)

## 方向性について

原田講師の下記の通り
> PostとTagは多数多の関係にして管理しているかと思うので、Postテーブルにtagカラムは必要ないかと思います。
うまくtagテーブルで管理できるよう処理を変更してみてください。

1 `def create`  で postは生成するがその際にstrong_parameters には tagは含めない
2 `def create`  で tagを別途 保存する (基礎編のbookmark編の様なイメージ)

form Rails 親子 あたりで検索すると良いと思いますー！

hidekazu-hhh commented:
> こちらのコードですが、正しくアソシエーションが設定されているのであれば (user側に)、 current_user.tags (現在のユーザーが投稿したタグ一覧) でとれるはずです。
> コントローラー (create) 時に、 デバッグし、 `ls current_user` などして current_userからtags メソッドが生えているか確認し、もしとなければ
> Userモデルを見直してみてください！ (おそらく当該のやりたいことはぐぐるとあるはず)

こちら、`current_user.tags`  で　ユーザーに紐づいたタグの一覧を取得できました。

コンソール

```
pry(main)> current_user = User.find(6)

pry(main)> **current_user.tags**
  Tag Load (0.6ms)  SELECT `tags`.* FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`user_id` = 6
```

コントローラー

```
@tag_user =  current_user.tags
```

次に**`each`** で　ユーザー自身に紐ついたタグの投稿一覧　を表示させるため、ビューの記述を変える必要があると思うのですが、下記の記述では他のユーザーの投稿も表示され期待した結果が得られません。
アソシエーションの設定に問題はないと思うので、
ビューの記述が誤っていると思います。

ビュー

```
<% @tag_user.each do |tag_user| %>
<%=tag_user.posts.map(&:content)%>
 <% end %>
```

```
 User.rb
  
  has_many :post_tags, dependent: :destroy
  has_many :tags,through: :post_tags
  has_many :posts, dependent: :destroy
  has_many :positive_words, dependent: :destroy

 Post.rb
 
  belongs_to :user
  has_many :post_tags, dependent: :destroy
  has_many :tags, through: :post_tags

 Tag.rb
 
  has_many :post_tags, dependent: :destroy, foreign_key: 'tag_id'
  has_many :posts, through: :post_tags
  has_many :users, through: :post_tags

 PostTag.rb

  belongs_to :post
  belongs_to :tag
  belongs_to :user

```





> ## 方向性について
> 原田講師の下記の通り
> 
> > PostとTagは多数多の関係にして管理しているかと思うので、Postテーブルにtagカラムは必要ないかと思います。
> > うまくtagテーブルで管理できるよう処理を変更してみてください。
> 
> 1 `def create` で postは生成するがその際にstrong_parameters には tagは含めない 2 `def create` で tagを別途 保存する (基礎編のbookmark編の様なイメージ)
> 
> form Rails 親子 あたりで検索すると良いと思いますー！
## 方向性について


Form Objectはデザインパターンの1つで、1つのフォームから複数のリソースを同時保存する際などに役立つ設計手法で本件に該当するのは理解しました。ですが、その他複数の記事を確認しましたが、
回答いただいた内容に沿った、実装のイメージが掴めていません。

[https://moneyforward.com/engineers_blog/2018/12/15/formobject/](https://moneyforward.com/engineers_blog/2018/12/15/formobject/)

[https://techtechmedia.com/form-object-implementation/](https://techtechmedia.com/form-object-implementation/)

1 `def create` で postは生成するがその際にstrong_parameters には tagは含めない

下記を参考にtag 用と post用にストロングパラメータをそれぞれ作成する。でしょうか？

[https://www.petitmonte.com/ruby/multiple_models.html](https://www.petitmonte.com/ruby/multiple_models.html)

2 `def create` で tagを別途 保存する (基礎編のbookmark編の様なイメージ)

### tagを別途 保存するとは、

postsコントローラーにtag を保存するロジックを記述する。

### 基礎編のbookmark編の様なイメージとは

中間テーブル用のコントローラーを新たに作成する。

でしょうか？

流れとしては

### ビュー

posts .new.htenl.erb　に`fields_for`　を記述　tag 　を作成。

### コントローラー

post コントローラー に　tag を作成するロジックを追加

### ****FormObjectの作成****

　　　　　　　　　　　　　でしょうか？

busitora commented:
できるだけ質問はフォーマットを整えて読み手のへ負荷を意識するともっと good です 👍 

1.  tag取得について
 > ビューの記述が誤っていると思います。他のユーザーの投稿も表示され期待した結果が得られません。

エラー文およびSQLを貼っていただかないとなんとも言えないですが(推測が難しい)、 `tag_user.posts.map(&:content)` をコンソールで1から設定してみて同じ、 `他のユーザーの投稿` になるか確認してみてください!!。もしならない場合はすべてこちらに貼っていただく＆仮説と検証結果も含めてお願いします。(tagに関しては同じ様な実装のアソシエーションの記事も相当数あると思いますので、そちらを元に仮説検証すれば解決出来る可能性は高いかと思われます)

2. ストロングパラメータについて
> def create で postは生成するがその際にstrong_parameters には tagは含めない
下記を参考にtag 用と post用にストロングパラメータをそれぞれ作成する。でしょうか？

それでも大丈夫です。ストロングパラメータで、 `params.require(:some_model)` に mergeメソッド等使うと同じことが出来ると思われますが、こちらはishiiさんにおまかせします。

3 中間テーブル用のコントローラについて
> postsコントローラーにtag を保存するロジックを記述する。

中間テーブル用コントローラー作らずとも、同じformからパラメータを送って同じメソッド内(別でも可)で作成できると思います。
`タグ機能 Rails` 等で 様々なやり方があると思うので、FormObject等にこだわらず柔軟にどうすれば実装できるか試行錯誤してみてください!!

PS 私は月木勤務なので、返信はどうしても遅れてしまいます...  なので、質問について再度あげる際は、スタテクメンバー(RUNTEQの 中井さん、廣瀬さん、土屋さんなど)にメンションを飛ばしたほうがスムーズかもしれません。(スタテクメンバー回答orその日勤務の講師に依頼 となるため)

hidekazu-hhh commented:
> エラー文およびSQLを貼っていただかないとなんとも言えないですが(推測が難しい)、 `tag_user.posts.map(&:content)` をコンソールで1から設定してみて同じ、 `他のユーザーの投稿` になるか確認してみてください!!。もしならない場合はすべてこちらに貼っていただく＆仮説と検証結果も含めてお願いします。(tagに関しては同じ様な実装のアソシエーションの記事も相当数あると思いますので、そちらを元に仮説検証すれば解決出来る可能性は高いかと思われます)

こちらについては検証します。ただ、当初、発生したエラーの原因がわかったのでこちらは、
解決済みとさせていただきます。ありがとうございました。

> `タグ機能 Rails` 等で 様々なやり方があると思うので、FormObject等にこだわらず柔軟にどうすれば実装できるか試行錯誤してみてください!!
> 
> PS 私は月木勤務なので、返信はどうしても遅れてしまいます... なので、質問について再度あげる際は、スタテクメンバー(RUNTEQの 中井さん、廣瀬さん、土屋さんなど)にメンションを飛ばしたほうがスムーズかもしれません。(スタテクメンバー回答orその日勤務の講師に依頼 となるため)

下記の記事の`accepts_nested_attributes_for`　を試したところ

[https://qiita.com/koki_73/items/bc4ca80ab43e84d9704f](https://qiita.com/koki_73/items/bc4ca80ab43e84d9704f)

`Unpermitted parameter: :tag` は解消しました。

こちら、検証用に作成したプロジェクトで試しました。（Userテーブルを除いたもの） 

post.rb

```
accepts_nested_attributes_for :tag_maps, allow_destroy: true
accepts_nested_attributes_for :tags
```

posts.コントローラ

```

def new
    @post = Post.new
    @tag = @post.tags.build
end

def create
    
    @post = Post.new(post_params)
    @post.save  
end

  private

  def post_params
    params.require(:post).permit(:title, :content, tags_attributes: [:name, :_destroy, :id])
  end
```

new.html.erb

```
<%= form_with model: @post do |post| %>
  
<%= post.fields_for :tags do |tag| %><br>
<%= tag.label :name %><br>
<%= tag.text_field :name %><br>
<%end%>
<%= post.submit %>
<% end %>
```

post.rb （save_tagsメソッド）

```
def save_tags(tags)
    
    tag_list = tags.split(/[[:blank:]]+/)

    current_tags = self.tags.pluck(:name)

    old_tags = current_tags - tag_list

    new_tags = tag_list - current_tags

    p current_tags

    old_tags.each do |old|
      self.tags.delete Tag.find_by(name: old)
    end

    new_tags.each do |new|
      new_post_tag = Tag.find_or_create_by(name: new)
      post_tags.new(user_id: user_id,tag_id: new_post_tag.id).save
      # self.tags << new_post_tag
    end
    
  end
end
```

ログ

```
Started POST "/posts" for ::1 at 2022-06-18 15:17:33 +0900
Processing by PostsController#create as JS
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"Z8+qjq08Eyd0zRimi8VE928X1/VgwxctnlVP/dCgTN79M6MqqihuaSuse9XgFiIVgbBf7GJUI9TyqCRNUkF3DA==", "post"=>{"title"=>"ss", "content"=>"ss", "tags_attributes"=>{"0"=>{"name"=>"水泳"}}}, "commit"=>"Create Post"}
   (0.1ms)  begin transaction
  ↳ app/controllers/posts_controller.rb:19
  Post Create (2.8ms)  INSERT INTO "posts" ("title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["title", "ss"], ["content", "ss"], ["created_at", "2022-06-18 06:17:33.242895"], ["updated_at", "2022-06-18 06:17:33.242895"]]
  ↳ app/controllers/posts_controller.rb:19
  Tag Create (0.2ms)  INSERT INTO "tags" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "水泳"], ["created_at", "2022-06-18 06:17:33.247574"], ["updated_at", "2022-06-18 06:17:33.247574"]]
  ↳ app/controllers/posts_controller.rb:19
  TagMap Create (0.5ms)  INSERT INTO "tag_maps" ("post_id", "tag_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["post_id", 56], ["tag_id", 43], ["created_at", "2022-06-18 06:17:33.257823"], ["updated_at", "2022-06-18 06:17:33.257823"]]
  ↳ app/controllers/posts_controller.rb:19
   (4.9ms)  commit transaction
  ↳ app/controllers/posts_controller.rb:19
Completed 500 Internal Server Error in 27ms (ActiveRecord: 8.7ms)

  
NoMethodError (undefined method `split' for #<ActionController::Parameters:0x00007fef8ed4de58>):
  
app/models/post.rb:18:in `save_tags'
app/controllers/posts_controller.rb:20:in `create'
```

post.rb　に定義した`post_save`メソッドを記述するとエラーになります。

ログ

```
NoMethodError (undefined method `split' for nil:NilClass):
```

デバック

```
[4] pry(#<PostsController>)> @post = Post.new(post_params)
=> #<Post:0x00007fabfe486708 id: nil, title: "あああ", content: "ssss", created_at: nil, updated_at: nil>
[5] pry(#<PostsController>)> @post.save 
   (0.1ms)  begin transaction
  ↳ (pry):5
  Post Create (2.1ms)  INSERT INTO "posts" ("title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["title", "あああ"], ["content", "ssss"], ["created_at", "2022-06-18 04:44:12.904561"], ["updated_at", "2022-06-18 04:44:12.904561"]]
  ↳ (pry):5
  Tag Create (0.6ms)  INSERT INTO "tags" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "英語"], ["created_at", "2022-06-18 04:44:12.908608"], ["updated_at", "2022-06-18 04:44:12.908608"]]
  ↳ (pry):5
  TagMap Create (1.0ms)  INSERT INTO "tag_maps" ("post_id", "tag_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["post_id", 55], ["tag_id", 42], ["created_at", "2022-06-18 04:44:12.935831"], ["updated_at", "2022-06-18 04:44:12.935831"]]
  ↳ (pry):5
   (4.2ms)  commit transaction
  ↳ (pry):5
=> true
[6] pry(#<PostsController>)> @post.save_tags 
ArgumentError: wrong number of arguments (given 0, expected 1)
from /Users/ishiihidekazu/learn_app/start_app/app/models/post.rb:14:in `save_tags'
[7] pry(#<PostsController>)> @post.save_tags(params[:post]) 
NoMethodError: undefined method `split' for #<ActionController::Parameters:0x00007fabfea83ac0>
```

これは、@post.save 時点で　

タグを保存しているため（new アクションでpostに紐づいたtag オブジェクトを作成 ）

`NoMethodError (undefined method `split' for nil:NilClass):`　になってしまう。

つまり、この方法では`post_save`メソッドを使う余地がないと思いました。

現状、`Unpermitted parameter: :tag` は解消しましたが、

重複したタグが作成される。タグに関連した投稿を取得するとき等に、期待した結果が得られない。少なくとも、この方法では`post_save`　メソッドを含めた記述ができないと思うので、他の方法を検証しておりますが、記事も断片的な内容で実装の方針が定まっておりません。

同じフォームにtag を含め　postコントローラ内で別途　tag を保存する方向性について、
ヒントをいただけないでしょうか？

takyyi commented:
save_tagsメソッドの実装を見る限り、accepts_nested_attributes_forの機能を使わなくてもよさそうな気がしますね。
一度、https://qiita.com/ki_87/items/a344ea566c88b10b950c#%E3%83%A2%E3%83%87%E3%83%AB%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB
などを参考にしてサンプルプロジェクトを作って試してみてはどうでしょうか？

hidekazu-hhh commented:
> save_tagsメソッドの実装を見る限り、accepts_nested_attributes_forの機能を使わなくてもよさそうな気がしますね。
> 一度、https://qiita.com/ki_87/items/a344ea566c88b10b950c#%E3%83%A2%E3%83%87%E3%83%AB%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB
> などを参考にしてサンプルプロジェクトを作って試してみてはどうでしょうか？


@takyyk 
こちら、上記を参考に作成しましたが、
ご回答の趣旨としてはサンプルプロジェクトを作り、それを元に色々と試してみては？
という事でしょうか？

takyyi commented:
@hidekazu-hhh 
そうですね。その趣旨であっています！

今回だと、```@post.save_tags(params[:post]) ```で```undefined method `split'```エラーが出ていますが、引数をうまく渡せていなさそうなので、上記リンクを参考にまず引数を適切に渡して、splitメソッドを使えるようにしてみてはということです！

hidekazu-hhh commented:
> そうですね。その趣旨であっています！
> 
> 今回だと、`@post.save_tags(params[:post]) `で`` undefined method `split' ``エラーが出ていますが、引数をうまく渡せていなさそうなので、上記リンクを参考にまず引数を適切に渡して、splitメソッドを使えるようにしてみてはということです！

@takyyk
別に作成したプロジェクトと比較した結果、
コントロールに記載したparams(引数名)とフォームに記述した内容 と一致していないため、`NoMethodError (undefined method `split' for nil:NilClass):`　が起きることを確認しました。

ログ　エラーのあるプロジェクト

```
[1] pry(#<PostsController>)> params
=> <ActionController::Parameters {"utf8"=>"✓", "authenticity_token"=>"9eJWi0vlblDyJyo6lG2jFrvENUqpU8uG7VKHl4c3wENvHl8vTPETHq1GSUn/vsX0VWO9U6vE/3+Br+wnBdb7kQ==", "post"=><ActionController::Parameters {"title"=>"assd", "content"=>"sっd", "tags_attributes"=>{"0"=>{"name"=>"ランニング"}}} permitted: false>, "commit"=>"Create Post", "controller"=>"posts", "action"=>"create"} permitted: false>
[2] pry(#<PostsController>)> params[:post][:name]
=> nil
[3] pry(#<PostsController>)> params[:post][:tags_attributes] 
=> <ActionController::Parameters {"0"=>{"name"=>"ランニング"}} permitted: false>
```

ログ　正常のプロジェクト

```
[1] pry(#<PostsController>)> params
=> <ActionController::Parameters {"utf8"=>"✓", "authenticity_token"=>"BRSjvSSNik8u+juC5x3Hb4XLtU6Jbn0gY2g9zgBT5/dnhylKroZdso3ugnnwM4YVYKYIpKdvaLDjT90FxO2YHA==", "post"=><ActionController::Parameters {"title"=>"efew", "content"=>"efef", "name"=>"旅行"} permitted: false>, "commit"=>"Create Post", "controller"=>"posts", "action"=>"create"} permitted: false>
[2] pry(#<PostsController>)> params[:post]
=> <ActionController::Parameters {"title"=>"efew", "content"=>"efef", "name"=>"旅行"} permitted: false>
[3] pry(#<PostsController>)> params[:post][:name] 
=> "旅行"
```

コントローラーの記述を**`:**name`に変更。

```
def create
    # パラメーターを受け取り保存準備
    @post = Post.new(post_params)
    if @post.save
        @post.save_tags(params[:post][:name])
        # 成功したらトップページへリダイレクト
     redirect_to root_path
    end
```

ログ  

```
Started POST "/posts" for ::1 at 2022-06-19 15:10:53 +0900
Processing by PostsController#create as JS
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"d+6o1pueET1nztl5aC11oLdcAGyInOd8fVuW8qEp6I6svQLJehHQmmVG/3CQp+mIYmZpgKydTzLGjGQlmhMt1g==", "post"=>{"title"=>"xvxz", "content"=>"zxvzxx", "tags_attributes"=>{"0"=>{"name"=>"英語"}}}, "commit"=>"Create Post"}
   (0.1ms)  begin transaction
  ↳ app/controllers/posts_controller.rb:19
  Post Create (0.6ms)  INSERT INTO "posts" ("title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["title", "xvxz"], ["content", "zxvzxx"], ["created_at", "2022-06-19 06:10:53.446564"], ["updated_at", "2022-06-19 06:10:53.446564"]]
  ↳ app/controllers/posts_controller.rb:19
  Tag Create (0.2ms)  INSERT INTO "tags" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "英語"], ["created_at", "2022-06-19 06:10:53.448457"], ["updated_at", "2022-06-19 06:10:53.448457"]]
  ↳ app/controllers/posts_controller.rb:19
  TagMap Create (0.4ms)  INSERT INTO "tag_maps" ("post_id", "tag_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["post_id", 137], ["tag_id", 124], ["created_at", "2022-06-19 06:10:53.458113"], ["updated_at", "2022-06-19 06:10:53.458113"]]
  ↳ app/controllers/posts_controller.rb:19
   (1.5ms)  commit transaction
  ↳ app/controllers/posts_controller.rb:19
Completed 500 Internal Server Error in 19ms (ActiveRecord: 3.0ms)

  
NoMethodError (undefined method `split' for nil:NilClass):
  
app/models/post.rb:18:in `save_tags'
app/controllers/posts_controller.rb:20:in `create'
```

`NoMethodError (undefined method `split' for nil:NilClass):` は変わらず。

### 不明点

色々と試して混乱してしまいました。

現時点でコントローラー側とビューの記述はこれで正しいと思っているのですが、

 コンソールで　`params[:post][:tags_attributes]` 　とすると

=> <ActionController::Parameters {"0"=>{"name"=>"ランニング"}} が返ってくるのでどちらが正しいのか混乱しています。

コントローラー側とビューの記述がこれで正しいのなら、

post.rb の記述　を修正する必要があると思っています。

```
def save_tags(tags)
    tag_list = tags.split(/[[:blank:]]+/)
```

 tag_list = tags.split(/[[:blank:]]+/)

`= tags.`　には　タグの名前が入っていてsplitで区切り配列にしていると思うのですが、

そもそも、この`= tags`　　によって、なぜタグの名前が渡っているのかがよく分かっていません。

ポートフォリオのプロジェクトを確認しましたが、分かっていません。

ポートフォリオのプロジェクト

フォーム

```
<%= post.label :tag, 'タグ' %><br>
<%= post.text_field :tag, value: @post.tags.pluck(:name) %><br>
```

 post コントローラー

```
def create
    # パラメーターを受け取り保存準備
    @post = current_user.posts.new(post_params)   
    if @post.save
      # タグの保存
      @post.save_tags(params[:post][:tag])
      # 成功したらトップページへリダイレクト
      redirect_to posts_path
    else
      # 失敗した場合は、newへ戻る
      render :new
    end
  end
```
htnl
<img width="710" alt="172b047dd9de66886f2394cd936177a6" src="https://user-images.githubusercontent.com/75820973/174470026-644a9f14-5f23-40b7-8c65-e8c8a246f81f.png">

takyyi commented:
@hidekazu-hhh 

> そもそも、この= tags　　によって、なぜタグの名前が渡っているのかがよく分かっていません。

これは```save_tags(tags)```メソッドの引数tagsであって、save_tags(params[:post][:name])メソッドが使われているところの引数にtagの名前がパラメータとして入って渡されていますね。

なので、どういう今回自分のプロジェクトではparameterでtagの名前が取れるのかデバッグして調べてみるというのをやってみたら良いかと思います。
今回、```accepts_nested_attributes_for```を使っているので、```"tags_attributes"=>{"0"=>{"name"=>"ランニング"}}}```のような形になっているかと思いますが、使わなければ、サンプルプロジェクトのような形のパラメータで進められるかと思います。

hidekazu-hhh commented:
> これは`save_tags(tags)`メソッドの引数tagsであって、save_tags(params[:post][:name])メソッドが使われているところの引数にtagの名前がパラメータとして入って渡されていますね。
> 
> なので、どういう今回自分のプロジェクトではparameterでtagの名前が取れるのかデバッグして調べてみるというのをやってみたら良いかと思います。

@takyyk
params[:post][:tag]　＝　save_tags(tags)　という意味だと理解しましたがあっていますでしょうか？

こちら確認しました。
デバック

```
[1] pry(#<PostsController>)> params[:post][:tag]
=> "プログラミング"
```

> 今回、`accepts_nested_attributes_for`を使っているので、`"tags_attributes"=>{"0"=>{"name"=>"ランニング"}}}`のような形になっているかと思いますが、使わなければ、サンプルプロジェクトのような形のパラメータで進められるかと思います。

つまり、accepts_nested_attributes_for`を使うなら別の記述が必要になる。
ということでしょうか？

takyyi commented:
@hidekazu-hhh 
> params[:post][:tag]　＝　save_tags(tags)　という意味だと理解しましたがあっていますでしょうか？

いえ、仮引数や実引数のようなワードで調べてみてください！引数を引き渡しているかそれとも受け取っているかという話です。

自分のプロジェクトで```params[:post][:tag]```にtag名が入っているのなら、それを引数としてtagsで受け取ることで、```splitメソッド```は使えるようになるかと思います。

そうすると```tags_attributes```のparameterは使っていないことになるという話です。

accepts_nested_attributes_forを使う場合はhttps://qiita.com/koki_73/items/bc4ca80ab43e84d9704fで例としてあげられているようなstrong parameterの設定だったりが必要になってきます。

hidekazu-hhh commented:
> いえ、仮引数や実引数のようなワードで調べてみてください！引数を引き渡しているかそれとも受け取っているかという話です。
> 
> 自分のプロジェクトで`params[:post][:tag]`にtag名が入っているのなら、それを引数としてtagsで受け取ることで、`splitメソッド`は使えるようになるかと思います。
> 
> そうすると`tags_attributes`のparameterは使っていないことになるという話です。
> 
> accepts_nested_attributes_forを使う場合は[https://qiita.com/koki_73/items/bc4ca80ab43e84d9704fで例としてあげられているようなstrong](https://qiita.com/koki_73/items/bc4ca80ab43e84d9704f%E3%81%A7%E4%BE%8B%E3%81%A8%E3%81%97%E3%81%A6%E3%81%82%E3%81%92%E3%82%89%E3%82%8C%E3%81%A6%E3%81%84%E3%82%8B%E3%82%88%E3%81%86%E3%81%AAstrong) parameterの設定だったりが必要になってきます。

こちら、`tags_attributes`　として　def create　に　

`@post.save_tags(params[:post][:tags_attributes]["0"][:name])`　

を記述し name の値　を取得。

```
[1] pry(#<PostsController>)> params[:post][:tags_attributes]["0"][:name]
=> "ランニング"
```

post.controller

```
def new
    @post = Post.new
    @tag = @post.tags.build
end

def create
    @post = Post.new(post_params)
   if @post.save 
        @post.save_tags(params[:post][:tags_attributes]["0"][:name])
        
     redirect_to root_path
   end
```

post.new.html.erb

```
<%= form_with model: @post do |post| %>
 <%= post.fields_for :tags do |tag| %><br>
<%= tag.label :name %><br>
<%= tag.text_field :name %><br>
<small>スペースで区切ることで複数指定できます。</small><br>
<%end%>
<%= post.submit %>
<%end%>
```

post.rb

```
def save_tags(tags)

    
    tag_list = tags.split(/[[:blank:]]+/)
    current_tags = self.tags.pluck(:name)

    old_tags = current_tags - tag_list

   
    new_tags = tag_list - current_tags

    p current_tags

    
    old_tags.each do |old|
    
      self.tags.delete Tag.find_by(name: old)
    end

  
    new_tags.each do |new|
      new_post_tag = Tag.find_or_create_by(name: new)
      self.tags << new_post_tag
    end

  end
```

ログ

```
Started POST "/posts" for ::1 at 2022-06-22 17:32:01 +0900
Processing by PostsController#create as JS
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"k2IKurmLHHa4H0IltNQs32n6/5T9pUnWQSAOrJcNCooJngMevp9hOOd+IVbfB0o9h113jf8yfS8t3WUcFewxWA==", "post"=>{"title"=>"あああ", "content"=>"dd", "tags_attributes"=>{"0"=>{"name"=>"ランニング"}}}, "commit"=>"Create Post"}
   (0.1ms)  begin transaction
  ↳ app/controllers/posts_controller.rb:19
  Post Create (0.6ms)  INSERT INTO "posts" ("title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["title", "あああ"], ["content", "dd"], ["created_at", "2022-06-22 08:32:01.409825"], ["updated_at", "2022-06-22 08:32:01.409825"]]
  ↳ app/controllers/posts_controller.rb:19
  Tag Create (0.2ms)  INSERT INTO "tags" ("name", "created_at", "updated_at") VALUES (?, ?, ?)  [["name", "ランニング"], ["created_at", "2022-06-22 08:32:01.412900"], ["updated_at", "2022-06-22 08:32:01.412900"]]
  ↳ app/controllers/posts_controller.rb:19
  TagMap Create (0.4ms)  INSERT INTO "tag_maps" ("post_id", "tag_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["post_id", 237], ["tag_id", 204], ["created_at", "2022-06-22 08:32:01.416363"], ["updated_at", "2022-06-22 08:32:01.416363"]]
  ↳ app/controllers/posts_controller.rb:19
   (2.1ms)  commit transaction
  ↳ app/controllers/posts_controller.rb:19
   (0.2ms)  SELECT "tags"."name" FROM "tags" INNER JOIN "tag_maps" ON "tags"."id" = "tag_maps"."tag_id" WHERE "tag_maps"."post_id" = ?  [["post_id", 237]]
  ↳ app/models/post.rb:20
["ランニング"]
Redirected to http://localhost:3000/
Completed 200 OK in 17ms (ActiveRecord: 3.5ms)
```

現状のエラー　

`NoMethodError (undefined method `split' for nil:NilClass)`　は解消しましたが、

save_tagsメソッドが動作せず、重複したタグが作成され、tags テーブルに保存されてしまう。

デバック

```
1] pry(#<PostsController>)> @post.save
   (0.1ms)  begin transaction
  ↳ (pry):1
   (0.1ms)  commit transaction
  ↳ (pry):1
=> true
[2] pry(#<PostsController>)> @post.save_tags(params[:post][:tags_attributes]["0"][:name]) 
   (0.4ms)  SELECT "tags"."name" FROM "tags" INNER JOIN "tag_maps" ON "tags"."id" = "tag_maps"."tag_id" WHERE "tag_maps"."post_id" = ?  [["post_id", 252]]
  ↳ app/models/post.rb:21
["ランニング"]
=> []
```

post.rb　の

 `tag_list = tags.split(/[[:blank:]]+/)`　に値が渡っているか確認。

```
[1] pry(#<Post>)> tag_list
=> ["ランニング"]
```

`tag_list`　には　値が代入されていますが、save_tagsメソッドが動作せず、

重複したタグが作成されてしまいます。

恐らく、上手く値が渡っていないからだと思いますが、どう対処すればよいかわからず詰まっています。

```
   @post.save_tags(params[:post][:tags_attributes]["0"][:name])
```

プルリクに上げているコードのログ

ログ

```
Started POST "/posts" for ::1 at 2022-06-23 17:24:25 +0900
Processing by PostsController#create as JS
  Parameters: {"authenticity_token"=>"8FJF5aCOkifcPYfKAAOy7ch70HIABINMEOj3mX2rDT4OvBEL4d55CZ3/iBk9Q4QqbrP/zzquNGPWT86TqVObNA==", "post"=>{"content"=>"あどみん太郎", "start_time(1i)"=>"2022", "start_time(2i)"=>"6", "start_time(3i)"=>"23", "tag"=>"テニス"}, "commit"=>"登録する"}
  User Load (1.2ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 6 LIMIT 1
Unpermitted parameter: :tag
   (0.9ms)  BEGIN
  ↳ app/controllers/posts_controller.rb:19:in `create'
  Post Create (1.9ms)  INSERT INTO `posts` (`content`, `start_time`, `user_id`, `created_at`, `updated_at`) VALUES ('あどみん太郎', '2022-06-23 00:00:00', 6, '2022-06-23 17:24:25.555775', '2022-06-23 17:24:25.555775')
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (2.2ms)  COMMIT
  ↳ app/controllers/posts_controller.rb:19:in `create'
   (0.3ms)  SELECT `tags`.`name` FROM `tags` INNER JOIN `post_tags` ON `tags`.`id` = `post_tags`.`tag_id` WHERE `post_tags`.`post_id` = 199
  ↳ app/models/post.rb:18:in `save_tags'
[]
  Tag Load (0.5ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`name` = 'テニス' LIMIT 1
  ↳ app/models/post.rb:42:in `block in save_tags'
   (1.5ms)  BEGIN
  ↳ app/models/post.rb:42:in `block in save_tags'
  Tag Exists? (1.8ms)  SELECT 1 AS one FROM `tags` WHERE `tags`.`name` = BINARY 'テニス' LIMIT 1
  ↳ app/models/post.rb:42:in `block in save_tags'
  Tag Create (1.8ms)  INSERT INTO `tags` (`name`, `created_at`, `updated_at`) VALUES ('テニス', '2022-06-23 17:24:25.585921', '2022-06-23 17:24:25.585921')
  ↳ app/models/post.rb:42:in `block in save_tags'
   (0.5ms)  COMMIT
  ↳ app/models/post.rb:42:in `block in save_tags'
   (0.3ms)  BEGIN
  ↳ app/models/post.rb:43:in `block in save_tags'
  Tag Load (0.4ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`id` = 31 LIMIT 1
  ↳ app/models/post.rb:43:in `block in save_tags'
  User Load (0.3ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 6 LIMIT 1
  ↳ app/models/post.rb:43:in `block in save_tags'
  PostTag Create (0.6ms)  INSERT INTO `post_tags` (`post_id`, `tag_id`, `user_id`, `created_at`, `updated_at`) VALUES (199, 31, 6, '2022-06-23 17:24:25.607873', '2022-06-23 17:24:25.607873')
  ↳ app/models/post.rb:43:in `block in save_tags'
   (1.2ms)  COMMIT
  ↳ app/models/post.rb:43:in `block in save_tags'
Redirected to http://localhost:3000/posts
Completed 200 OK in 65ms (ActiveRecord: 16.2ms | Allocations: 25081)
```

save_tags メソッドが動作する。重複するタグがあるか検索。なければ タグ を作成し保存。　

中間テーブルも保存。

他に試したこと

tag と　post を分けて保存する方法。tag のストロングパラメータ側でエラー。

`Unpermitted parameters: :title, :content`　

```
def new
    @post = Post.new
    @tag = Tag.new
  end

  def create
    @post = Post.new(post_params)
    @tag = Tag.new(tag_params)
    if @post.save && @tag.save
        @post.save_tags(params[:post][:name])
     redirect_to root_path
    end
    
   end
  

  private

  def post_params
    params.require(:post).permit(:title, :content)
  end
  
  def tag_params
    params.require(:post).permit(:name)
  end
end
```

```
<%= form_for @post do |post| %>
 
<%= fields_for @tags do |tag| %><br>
  <%= tag.label :name %><br>
  <%= tag.text_field :name %><br>
  <%end%>
<%= post.submit %>
<% end %>
```

試したこと

下記、記事を参考に新たにサンプルプロジェクトを作成し　

ストロングパラメータを２つ作成する方法で試しましたが

[https://qiita.com/ryutaro9595/items/042a1ec713c8c1f2c1d6](https://qiita.com/ryutaro9595/items/042a1ec713c8c1f2c1d6)

ログ

`Unpermitted parameter: :tag_names`　変わらず。

```
Processing by ArticlesController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"HooddhrCRT3V4KRgnutWkfKpSJT1Wp6YP7E0ccYeaqFAm03FshZxaCGPeOFAOlSxskfMUFQXunP4RB4cdZz0sQ==", "article"=>{"title"=>"qwew", "tag_names"=>"dasd", "content"=>"adasdas"}, "commit"=>"投稿"}
Unpermitted parameter: :tag_names
   (0.2ms)  begin transaction
  ↳ app/controllers/articles_controller.rb:8
  Article Create (1.3ms)  INSERT INTO "articles" ("title", "content", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["title", "qwew"], ["content", "adasdas"], ["created_at", "2022-06-23 05:12:53.175561"], ["updated_at", "2022-06-23 05:12:53.175561"]]
  ↳ app/controllers/articles_controller.rb:8
   (0.9ms)  commit transaction
  ↳ app/controllers/articles_controller.rb:8
Unpermitted parameters: :title, :content
  Tag Load (0.5ms)  SELECT  "tags".* FROM "tags" WHERE "tags"."tag_name" = ? LIMIT ?  [["tag_name", "dasd"], ["LIMIT", 1]]
  ↳ app/models/article.rb:9
   (0.0ms)  begin transaction
  ↳ app/models/article.rb:14
  Tag Exists (0.1ms)  SELECT  1 AS one FROM "tags" WHERE "tags"."tag_name" = ? LIMIT ?  [["tag_name", "dasd"], ["LIMIT", 1]]
  ↳ app/models/article.rb:14
  Tag Create (0.7ms)  INSERT INTO "tags" ("tag_name", "created_at", "updated_at") VALUES (?, ?, ?)  [["tag_name", "dasd"], ["created_at", "2022-06-23 05:12:53.228758"], ["updated_at", "2022-06-23 05:12:53.228758"]]
  ↳ app/models/article.rb:14
  ArticleTagRelation Create (1.5ms)  INSERT INTO "article_tag_relations" ("article_id", "tag_id", "created_at", "updated_at") VALUES (?, ?, ?, ?)  [["article_id", 1], ["tag_id", 1], ["created_at", "2022-06-23 05:12:53.239379"], ["updated_at", "2022-06-23 05:12:53.239379"]]
  ↳ app/models/article.rb:14
   (0.8ms)  commit transaction
  ↳ app/models/article.rb:14
Redirected to http://localhost:3000/articles/1
Completed 302 Found in 71ms (ActiveRecord: 7.3ms)
```

`collection_check_boxes` の方法もありますが、私が想定している仕様と異なり参考にできません。

[https://qiita.com/YUD96/items/f751010f8a58efaef6ed](https://qiita.com/YUD96/items/f751010f8a58efaef6ed)

タグ機能の記事について何度も目を通しましたが、本件に関する内容の記事は見当たりませんでした。今回、`accepts_nested_attributes_for`　で　`Unpermitted parameter`　は解消しましたが、この方法はあまり使用しない方が良い。という記事もあり、他の方法を検討していますが、

調べた限りではformオブジェクトを用いた方法になると思います。ですが、以前、formオブジェクト以外にも方法はあるとアドバイスいただき、調べましたが、本件を解消する記事は見当たりませんでした。formオブジェクト以外の方法とは、どのような方法でしょうか？もう少し具体的なヒントをいただけないでしょうか。現状、実現したいことは`Unpermitted parameter`　　を解消し、save_tags メソッドを動作させたいです。

また、`Unpermitted parameter`　というエラーは出ていますが、期待した動作はしています。

セキュリティ上に問題あることは承知しておりますが、もし仮にこのまま進めた場合、何か不具合があるのでしょうか？（デプロイできない等）

murata0705 commented:
@hidekazu-hhh 

> また、Unpermitted parameter　というエラーは出ていますが、期待した動作はしています。
セキュリティ上に問題あることは承知しておりますが、もし仮にこのまま進めた場合、何か不具合があるのでしょうか？（デプロイできない等）

Unpermitted parameter エラーが出ているが問題ないのであれば、一旦大丈夫かなと思います。デプロイできない等の問題は発生しません。

```ruby
params.require(:post).permit(:title, :content)
```

このような書き方をして、params をホワイトリスト形式で指定することをストロングパラメータといいますが、ストロングパラメータは、マスアサインメント脆弱性に対応するために rails に導入されました。

- https://rubyonrails.org/2012/3/21/strong-parameters

日本語だとこの記事がわかりやすいはずです。

- https://qiita.com/tbpgr/items/63028f5e19a2d1617e40

要は、

```ruby
User.create(params)
```

みたいな書き方をすると、params には、`{ name: "test", age: 20 }` を想定していたのに、 `{ name: "test", age: 20, admin: true }` みたいなデータを仕込まれたときにやばいねって話です。
rails 4 以降はエラーが出るはずで、そもそも上記の書き方はできません。

基本的にストロングパラメータを利用したほうが良いですが、

```ruby
tag = params[:post][:tag]
```

というように情報を取得すれば、そもそもホワイトリストなので、問題ないかなと思います。

murata0705 commented:
問題ないと書きましたが、できれば上記エラーができないようにきちんとコードを書くべきだと思います。
ただそのためには、

- [ActiveRecord::NestedAttributes::ClassMethods](https://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html)
- [Action Controller Parameters](https://api.rubyonrails.org/classes/ActionController/Parameters.html)

あたりの理解を深める必要があり、時間がかかりそうなためです。
今回は、Unpermitted parameter エラーが出ていますが、特にセキュリティリスクはなさそうなので、大丈夫かなという意味です。

hidekazu-hhh commented:
利用規約などの作成
